### PR TITLE
Fixed #13910 -- Added generator based rendering (streaming) to templates and supporting views.

### PR DESCRIPTION
**Pull request by pglotov**
Rebased and optimized https://github.com/django/django/pull/4783

### CODE REVIEW

File: django/template/defaulttags.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 """Default tags used by the template system, available to all templates."""
-import re
-import sys
-import warnings
+from __future__ import unicode_literals
```
Comment (auvipy) [inline django/template/defaulttags.py]: is this intentionally added?

File: docs/releases/2.2.txt [hunk 2]
```diff
@@ -470,6 +470,16 @@ Miscellaneous
   with several third-party apps that had models in tests without migrations.
   You must add migrations for such models.
 
+* Added :class:`~django.template.response.StreamingTemplateResponse` and
```
Comment (auvipy) [inline docs/releases/2.2.txt]: this should be on django 3.0

File: django/template/defaulttags.py (hunk 1 above)
Comment (pglotov) [reply django/template/defaulttags.py]: im cleaning it up atm ..

File: docs/releases/2.2.txt (hunk 2 above)
Comment (pglotov) [reply docs/releases/2.2.txt]: hm .. ok

File: django/template/base.py [hunk 3]
```diff
@@ -988,8 +988,7 @@ def render(self, context, stream=False):
     def _stream(self, context, stream=True):
         for node in self:
             if isinstance(node, Node):
-                for bit in node.stream_annotated(context):
-                    yield mark_safe(str(bit)) if stream else str(bit)
+                yield from map(lambda bit: mark_safe(str(bit)) if stream else str(bit), node.stream_annotated(context))
```
Comment (Gagaro) [inline django/template/base.py]: Wouldn't `(mark_safe(str(bit)) if stream else str(bit) for bit in node.stream_annotated(context))` be better here? To avoid the map/lambda combo.

Comment (pglotov) [reply django/template/base.py]: made changes

**Comment by Gagaro** [comment]
Thanks for taking care of cleaning that up @pglotov :+1: . I'm thrilled to see that in 3.0 :smile: .

**Comment by pglotov** [comment]
Yes, it seemed easier than rigging something up to handle timing out requests :)

**Comment by pglotov** [comment]
I have no idea why **pr-mariadb/database=mysql,label=mariadb,python=python3.7** fails.

File: docs/howto/custom-template-tags.txt [hunk 4]
```diff
@@ -757,6 +757,24 @@ Notes:
...
+
+    class OneToNineNode(template.Node):
+        def render(self, context):
+            return ''.join(self.stream(context))
+
+        def stream(self, context):
+            for i in range(1, 10):
+                yield str(i)
+
+  .. versionadded:: 1.10
```
Comment (Gagaro) [inline docs/howto/custom-template-tags.txt]: 3.0

File: docs/ref/template-response.txt [hunk 5]
```diff
@@ -292,3 +292,43 @@ with a simple template and a context containing a queryset::
 
     def blog_index(request):
         return TemplateResponse(request, 'entry_list.html', {'entries': Entry.objects.all()})
+
+``StreamingTemplateResponse`` and ``SimpleStreamingTemplateResponse``
+=====================================================================
+
+.. versionadded:: 1.10
```
Comment (Gagaro) [inline docs/ref/template-response.txt]: 3.0

File: docs/ref/templates/api.txt [hunk 6]
```diff
@@ -224,6 +224,17 @@ different contexts.
         >>> template.render(context)
         "My name is Dolores."
 
+    .. versionadded:: 1.10
```
Comment (Gagaro) [inline docs/ref/templates/api.txt]: 3.0

File: docs/topics/http/shortcuts.txt [hunk 7]
```diff
@@ -56,6 +56,14 @@ Optional arguments
     The :setting:`NAME <TEMPLATES-NAME>` of a template engine to use for
     loading the template.
 
+``stream``
+   If set to True, a :class:`~django.http.StreamingHttpResponse` will be
+   return instead.
+
+   .. versionchanged:: 1.10
```
Comment (Gagaro) [inline docs/topics/http/shortcuts.txt]: 3.0

File: docs/topics/templates.txt [hunk 8]
```diff
@@ -169,6 +169,13 @@ must provide a ``render()`` method with the following signature:
     Then the engine must make it, as well as the CSRF token, available in the
     template. How this is achieved is up to each backend.
 
+    If ``stream`` is set to True, it will return a generator for rendering
+    the template instead
+
+    .. versionadded:: 1.10
```
Comment (Gagaro) [inline docs/topics/templates.txt]: 3.0

File: docs/topics/templates.txt [hunk 9]
```diff
@@ -263,11 +270,19 @@ templates, Django provides a shortcut function which automates the process.
         An optional template engine :setting:`NAME <TEMPLATES-NAME>`. The
         search for the template will be restricted to that engine.
 
+    ``stream``
+        If set to ``True``, it will return a generator of the template instead.
+
+        .. versionadded:: 2.2
```
Comment (Gagaro) [inline docs/topics/templates.txt]: 3.0

**Comment by Gagaro** [comment]
The failed test doesn't seems to be on our end: `ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?`

**Comment by auvipy** [comment]
it would need you to squash commits in a single one using co-authored with tag in git I guess.

**Comment by Gagaro** [comment]
Ok that should be all good then.

**Comment by Gagaro** [comment]
It's still red:

```
test_chunked_data (builtin_server.tests.ServerHandlerChunksProperly) ... Killed
Build step 'Execute shell' marked build as failure
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?
Finished: FAILURE
```

**Comment by pglotov** [comment]
Looks like several other recent PRs fail this test with same exact message.

**Comment by pglotov** [comment]
Performance: master vs pr:
```
Server Software:        WSGIServer/0.2
Server Hostname:        0.0.0.0
Server Port:            8001

Document Path:          /polls/streaming/
Document Length:        268906 bytes

Concurrency Level:      1
Time taken for tests:   16.633 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      26909300 bytes
HTML transferred:       26890600 bytes
Requests per second:    6.01 [#/sec] (mean)
Time per request:       166.333 [ms] (mean)
Time per request:       166.333 [ms] (mean, across all concurrent requests)
Transfer rate:          1579.88 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       0
Processing:   141  166  28.4    153     279
Waiting:      140  166  28.3    153     278
Total:        141  166  28.4    153     279

Percentage of the requests served within a certain time (ms)
  50%    153
  66%    163
  75%    172
  80%    189
  90%    207
  95%    232
  98%    262
  99%    279
 100%    279 (longest request)
######################################################
Server Software:        WSGIServer/0.2
Server Hostname:        0.0.0.0
Server Port:            8001

Document Path:          /polls/streaming/
Document Length:        268906 bytes

Concurrency Level:      1
Time taken for tests:   15.726 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      26909300 bytes
HTML transferred:       26890600 bytes
Requests per second:    6.36 [#/sec] (mean)
Time per request:       157.262 [ms] (mean)
Time per request:       157.262 [ms] (mean, across all concurrent requests)
Transfer rate:          1671.01 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:   140  157  15.4    153     217
Waiting:      140  157  15.4    153     216
Total:        140  157  15.4    153     217

Percentage of the requests served within a certain time (ms)
  50%    153
  66%    157
  75%    161
  80%    163
  90%    174
  95%    190
  98%    215
  99%    217
 100%    217 (longest request)
```

**Comment by Gagaro** [comment]
What are the next steps to see this PR merged ?

**Comment by auvipy** [comment]
I guess uncheck the button patch needs improvement and need docs test etc in the related ticket so that the core fellows could do the final round of review to merge it.

**Comment by pglotov** [comment]
I don't think i have permissions to uncheck the flag, is anyone able to? https://code.djangoproject.com/ticket/13910

**Comment by auvipy** [comment]
done

**Comment by Gagaro** [comment]
Thanks :+1:

File: django/template/base.py [hunk 10]
```diff
@@ -718,6 +740,7 @@ def args_check(name, func, provided):
                                       (name, alen - dlen, plen))
 
         return True
+
```
Comment (felixxm) [inline django/template/base.py]: Please revert all unrelated changes e.g. new blank lines in various files.

File: django/template/base.py [hunk 11]
```diff
@@ -843,7 +866,7 @@ def _resolve_lookup(self, context):
                             current = current[int(bit)]
                         except (IndexError,  # list index out of range
                                 ValueError,  # invalid literal for int()
-                                KeyError,    # current is a dict without `int(bit)` key
+                                KeyError,  # current is a dict without `int(bit)` key
```
Comment (felixxm) [inline django/template/base.py]: Please revert this change.

**Comment by pglotov** [comment]
done

File: django/template/base.py [hunk 12]
```diff
@@ -880,6 +895,20 @@ def _resolve_lookup(self, context):
...
+    """
+    Check for errors during template rendering. If debug is True, contextual
+    line information is added to the exception before it is reraised.
+    """
+    try:
+        yield
+    except Exception as e:
+        if context.template.engine.debug and not hasattr(e, 'template_debug'):
+            e.template_debug = context.render_context.template.get_exception_info(e, token)
+        raise
```
Comment (felixxm) [inline django/template/base.py]: I would moved `annotate_exception` context manager to the `Node` class.

File: django/template/base.py [hunk 13]
```diff
@@ -880,6 +895,20 @@ def _resolve_lookup(self, context):
         return current
 
 
+@contextmanager
+def annotate_exception(context, token):
+    """
+    Check for errors during template rendering. If debug is True, contextual
+    line information is added to the exception before it is reraised.
```
Comment (felixxm) [inline django/template/base.py]:
```
"""     
If DEBUG is True, add contextual line information to exceptions
occurred in the template during rendering.
"""
```

File: django/template/backends/dummy.py [hunk 14]
```diff
@@ -42,7 +42,10 @@ def get_template(self, template_name):
 
 class Template(string.Template):
 
-    def render(self, context=None, request=None):
+    def render(self, context=None, request=None, stream=False):
+        if stream:
+            raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/template/backends/dummy.py]: Test is missing.

**Comment by pglotov** [comment]
made changes

File: django/template/backends/dummy.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pglotov) [reply django/template/backends/dummy.py]: added a test

**Comment by carltongibson** [comment]
For reference, a link @Gagaro's benchmarking project for this change. https://github.com/Gagaro/django_benchmark_streaming_templates

**Comment by pglotov** [comment]
Hi @carltongibson, i'm not sure it's a good idea to only leave `Template.render(... stream=True)` as the only way to use template streaming. I was actually looking forward to just replace `TemplateView` with `StreamingTemplateView` as base class in a bunch of views which tend to cause timeout on aws.

**Comment by carltongibson** [comment]
Hi @pglotov. Sure. You're welcome to include versions of SreamingTemplateResponse and StreamingTemplateView in your own project.

**Comment by pglotov** [comment]
Hmm.. isn't the point of Django to make users' life easier?

**Comment by pglotov** [comment]
It definitely would NOT make my life easier .. I would like to get a second opinion on this. How should i go about it (I'm not familiar with power hierarchy in Django project).

**Comment by pglotov** [comment]
I removed `StreamingTemplateView` related classes, and I'd like to find a solution (possibly adding `stream=True` parameter to TemplateView) to be able to easily use template streaming with a class based view.

File: django/template/loader.py [hunk 15]
```diff
@@ -49,7 +49,7 @@ def select_template(template_name_list, using=None):
         raise TemplateDoesNotExist("No template names provided")
 
 
-def render_to_string(template_name, context=None, request=None, using=None):
+def render_to_string(template_name, context=None, request=None, using=None, stream=False):
```
Comment (carltongibson) [inline django/template/loader.py]: Please remove.

File: django/template/engine.py [hunk 16]
```diff
@@ -146,7 +146,7 @@ def get_template(self, template_name):
             template = Template(template, origin, template_name, engine=self)
         return template
 
-    def render_to_string(self, template_name, context=None):
+    def render_to_string(self, template_name, context=None, stream=False):
```
Comment (carltongibson) [inline django/template/engine.py]: Please remove.

File: docs/internals/deprecation.txt [hunk 17]
```diff
@@ -444,6 +444,7 @@ details on these changes.
   * ``django.template.Context()``
   * ``django.template.RequestContext()``
   * ``django.template.response.TemplateResponse()``
+  * ``django.template.response.StreamingTemplateResponse()``
```
Comment (carltongibson) [inline docs/internals/deprecation.txt]: Please remove.

File: docs/ref/template-response.txt [hunk 18]
```diff
@@ -292,3 +292,43 @@ with a simple template and a context containing a queryset::
 
     def blog_index(request):
         return TemplateResponse(request, 'entry_list.html', {'entries': Entry.objects.all()})
+
+``StreamingTemplateResponse`` and ``SimpleStreamingTemplateResponse``
```
Comment (carltongibson) [inline docs/ref/template-response.txt]: Please remove.

File: tests/template_tests/syntax_tests/test_if.py [hunk 19]
```diff
@@ -601,6 +601,12 @@ def test_if_is_not_both_variables_missing(self):
         output = self.engine.render_to_string('template', {})
         self.assertEqual(output, 'no')
 
+    @setup({'if-tag-stream': '{% if foo %}{{ a }}42{% else %}no{% endif %}'})
+    def test_if_tag_stream(self):
+        output = self.engine.render_to_string('if-tag-stream', {'foo': True, 'a': 21}, stream=True)
+        self.assertEqual(next(output), '21')
+        self.assertEqual(next(output), '42')
```
Comment (felixxm) [inline tests/template_tests/syntax_tests/test_if.py]: Please remove this and other tests related with adding `stream` argument to the `render_to_string()` and revert also docs changes.

**Comment by pglotov** [comment]
made changes

**Comment by pope1ni** [comment]
@carltongibson @felixxm We seem to have a problem with pull requests for older Django Trac tickets closing unrelated open pull requests, e.g. this one.

This is because we structure commit messages with the prefix `Fixed #11157 -- ...` which GitHub takes to mean that issue or PR #11157 is resolved by that commit. See [here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

I fear that this will become more common. Is it time to consider changing the format of commit messages, e.g. `Fixed T11157 -- ...` or `Fixed ticket-11157 -- ...`? (Or something better than those...)

**Comment by carltongibson** [comment]
Hey @pope1ni. Yes, it happens from time to time. 🤔

Let's close due to inactivity. Great if someone wants to pick this up.
